### PR TITLE
Added support for ErgoDox with STM32 Microcontroller

### PR DESCRIPTION
**Pull request by Codetector1374**
## Description
Added support for Ergodox with stm32 microcontroller.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/ergodox_stm32/ergodox_stm32.c [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+    matrix_init_user();
+}
+
+void ergodox_blink_all_leds(void)
+{
+    ergodox_led_all_off();
+    // ergodox_led_all_set(LED_BRIGHTNESS_DEFAULT);
+    ergodox_board_led_1_on();
+    chThdSleepMilliseconds(50);
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.c]:
```suggestion
    wait_ms(50);
```

File: keyboards/ergodox_stm32/ergodox_stm32.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    ergodox_board_led_2_on();
+    chThdSleepMilliseconds(50);
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.c]:
```suggestion
    wait_ms(50);
```

File: keyboards/ergodox_stm32/ergodox_stm32.c [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    ergodox_board_led_3_on();
+    chThdSleepMilliseconds(50);
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.c]:
```suggestion
    wait_ms(50);
```

File: keyboards/ergodox_stm32/ergodox_stm32.c [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    ergodox_board_led_1_off();
+    chThdSleepMilliseconds(50);
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.c]:
```suggestion
    wait_ms(50);
```

File: keyboards/ergodox_stm32/ergodox_stm32.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    ergodox_board_led_2_off();
+    chThdSleepMilliseconds(50);
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.c]:
```suggestion
    wait_ms(50);
```

File: keyboards/ergodox_stm32/config.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifndef CONFIG_H
```
Comment (drashna) [inline keyboards/ergodox_stm32/config.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

File: keyboards/ergodox_stm32/config.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+#define CONFIG_H
```
Comment (drashna) [inline keyboards/ergodox_stm32/config.h]:
```suggestion
```

File: keyboards/ergodox_stm32/config.h [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+#define MATRIX_ROWS 14
+#define MATRIX_ROWS_PER_SIDE (MATRIX_ROWS / 2)
+#define MATRIX_COLS 6
+/* key combination for command */
+#define IS_COMMAND() ( \
+    keyboard_report->mods == (MOD_BIT(KC_LCTL) | MOD_BIT(KC_RCTL)) || \
+    keyboard_report->mods == (MOD_BIT(KC_LSFT) | MOD_BIT(KC_RSFT)) \
+)
+#endif
```
Comment (drashna) [inline keyboards/ergodox_stm32/config.h]:
```suggestion
```

File: keyboards/ergodox_stm32/ergodox_stm32.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
+#ifndef ERGODOX_STM32_H
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.h]: If you could replace the include guard (the ifndef, define, and endif at the end) with just `#pragma once`? 

It's simpler, and less prone to user error, so if you could make this switch? 

```suggestion
#pragma once
```

File: keyboards/ergodox_stm32/ergodox_stm32.h [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+#define ERGODOX_STM32_H
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.h]:
```suggestion
```

File: keyboards/ergodox_stm32/ergodox_stm32.h [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        {R01, R11, R21, R31, KC_NO, R51},                                 \
+        {R02, R12, R22, R32, R42, R52},                                   \
+        {R03, R13, R23, R33, R43, R53},                                   \
+        {R04, R14, R24, R34, R44, R54},                                   \
+        {R05, R15, R25, R35, R45, R55},                                   \
+    {                                                                     \
+      R06, R16, R26, R36, R46, KC_NO                                      \
+    }                                                                     \
+  }
+#endif
```
Comment (drashna) [inline keyboards/ergodox_stm32/ergodox_stm32.h]:
```suggestion
```

File: keyboards/ergodox_stm32/i2cmaster.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
+/* Copyright 2018 Jack Humbert
```
Comment (drashna) [inline keyboards/ergodox_stm32/i2cmaster.h]: This file isn't needed, and should be removed

File: keyboards/ergodox_stm32/rules.mk [hunk 13]
```diff
@@ -0,0 +1,30 @@
+SRC += matrix.c \
+i2cmaster.c
```
Comment (drashna) [inline keyboards/ergodox_stm32/rules.mk]:
```suggestion
    i2c_master.c
```

File: keyboards/ergodox_stm32/i2cmaster.c [hunk 14]
```diff
@@ -0,0 +1,101 @@
+/* Copyright 2018 Jack Humbert
```
Comment (drashna) [inline keyboards/ergodox_stm32/i2cmaster.c]: Please remove this file. 

Specifically, we want to use `/drivers/arm/i2c_master.c` instead of a custom implementation.

Comment (Codetector1374) [reply keyboards/ergodox_stm32/i2cmaster.c]: I guess I should modify /driver/ implementation to work with STM32 F1 devices then?

**Comment by yanfali** [comment]
@Codetector1374 if you get a chance to add info.json that would be helpful for configurator.

File: keyboards/ergodox_stm32/i2cmaster.c (hunk 14 above)
Comment (drashna) [reply keyboards/ergodox_stm32/i2cmaster.c]: In theory, it should already.

Comment (drashna) [reply keyboards/ergodox_stm32/i2cmaster.c]: In fact, it looks almost identical to the existing implementation, except for some hard coded behavior. 

So you really shouldn't need to do much here, at all.

Comment (Codetector1374) [reply keyboards/ergodox_stm32/i2cmaster.c]: Oh so the chibiOS API has 2 different initialization structures for F1 and F4 etc.. i2C the initlization part needs to change. And that’s all. Both the GPIO and i2C config struct. I am thinking about doing ifdefs for that. Detect F1 family device.

**Comment by Codetector1374** [comment]
I have addressed all the changes above.

**Comment by yanfali** [comment]
I've tested info.json on config preview. Looks good. Thank you. If you want to submit a default keymap to configurator send me a PR. Thanks

File: drivers/arm/i2c_master.c [hunk 15]
```diff
@@ -32,6 +32,15 @@
 
 static uint8_t i2c_address;
 
+// ChibiOS uses two initialization structure for v1 and v2/v3 i2c APIs.
```
Comment (drashna) [inline drivers/arm/i2c_master.c]: @yiancar since I think you're more familiar with this, any issues with this?

Comment (Codetector1374) [reply drivers/arm/i2c_master.c]: Is there anything else I should improve about this? Or is this good to merge?

Comment (yiancar) [reply drivers/arm/i2c_master.c]: I actually have never checked what API the f1 uses. Just to make sure @awkannan any takes on this?

Comment (pelrun) [reply drivers/arm/i2c_master.c]: I'd prefer more general means of determining the i2c code to use rather than a specific MCU symbol definition, although ChibiOS doesn't exactly make it easy. Even if it was just the generic family definition instead (`STM32F1xx`) I'd be happy enough.

Comment (awkannan) [reply drivers/arm/i2c_master.c]: I haven't tested this i2c code on STM32F1xx.

I agree with @pelrun. There are also multiple i2c buses available to use on STM chips, so this still is not enough. 

Until we figure out a good way to handle i2c bus selection and using family definition, I'd personally prefer to keep this in keyboard branches, etc. Just override i2c_master in your local keyboard folder. But I'm not the judge of that, I defer to QMK people to figure it out. The reason I say this though is because my RGB code is stuck in purgatory because it doesn't support every STM32 chip yet.

Also, as it stands, any non STM32F1xx chip would fall back to the F303 code, which is wrong on many chips. I like how we do it in eeprom, where we raise an error if no one has written the code for a specific family.

I also don't know how I feel about making tons of ifdefs, but I don't know a better alternative.

Comment (pelrun) [reply drivers/arm/i2c_master.c]: The best solution would probably be to add an explicit symbol in the board definition and choose based on that. I was planning on bringing in to core the main board definitions that are floating around some of the keyboards, at which point it's fairly straightforward to keep them in line.

Comment (Codetector1374) [reply drivers/arm/i2c_master.c]: I agree with @pelrun, however I didn't make it so that you have to declare this in board.h because I feel like that would break compatibility with a lot of existing code. I am thinking about either let the board.h define the config structure and initialization, thus they can both control the i2c version and which i2c bus to use. However, again, this would come at the cost of backwards compatibility. I initially just provided my own i2c_master file, but @drashna suggested to me that I should probably use the standard i2c_master. So I modified the i2c master.

Comment (drashna) [reply drivers/arm/i2c_master.c]: If this is an issue right now, then it should be fine to revert back to the separate files.  

Though, I'm not fond of having multiple files, but it seems like this is the best option until we have a better way to handle this.

File: .vscode/settings.json [hunk 16]
```diff
@@ -12,6 +12,10 @@
         "*.c": "c",
         "*.cpp": "cpp",
         "*.hpp": "cpp",
-        "xstddef": "c"
+        "xstddef": "c",
```
Comment (skullydazed) [inline .vscode/settings.json]: Can you pull these out? If these are good changes we can accept them in another PR, but we don't like to group unrelated changes in.

File: keyboards/ergodox_stm32/keymaps/default/keymap.c [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+  [2] = LAYOUT_ergodox(KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_NUMLOCK,KC_KP_SLASH,KC_KP_ASTERISK,KC_KP_MINUS,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_KP_7,KC_KP_8,KC_KP_9,KC_KP_PLUS,KC_TRANSPARENT,KC_TRANSPARENT,KC_KP_4,KC_KP_5,KC_KP_6,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_KP_1,KC_KP_2,KC_KP_3,KC_ENTER,KC_TRANSPARENT,KC_KP_DOT,KC_KP_0,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT,KC_TRANSPARENT),
+
+};
+
+const uint16_t PROGMEM fn_actions[] = {
```
Comment (skullydazed) [inline keyboards/ergodox_stm32/keymaps/default/keymap.c]: Instead of using fn_actions, which we deprecated, could you use one of the quantum keys described here? https://docs.qmk.fm/#/feature_advanced_keycodes?id=switching-and-toggling-layers

I think in this case you want they keycode `TT(1)`.

File: keyboards/ergodox_stm32/matrix.c [hunk 18]
```diff
@@ -0,0 +1,196 @@
+#include <stdint.h>
```
Comment (drashna) [inline keyboards/ergodox_stm32/matrix.c]: I don't know how much of this code is based on the Ergodox EZ code, sinc I haven't taken a close look side by side here ....  

But there have been some significant changes to the ergodox ez's custom matrix to allow for better debouncing.  You may want to review those changes and implement them here, as well.   Though, I would recommend the default debounce type.

File: keyboards/ergodox_stm32/rules.mk [hunk 19]
```diff
@@ -0,0 +1,30 @@
+SRC += matrix.c \
+drivers/arm/i2c_master.c
```
Comment (drashna) [inline keyboards/ergodox_stm32/rules.mk]: You don't need the full path here, actually.  Specifically, the drivers folder is added to the VPATH (virtual path) for the project, so it should pick it up, without the full path. 
```suggestion
    i2c_master.c
```

**Comment by drashna** [comment]
There are some merge conflicts now. 

If you could update your repo, and fix those?

**Comment by Codetector1374** [comment]
ahh sure I will find sometime this week clear those conflicts

**Comment by drashna** [comment]
Awesome, thanks!

**Comment by Codetector1374** [comment]
Okay finally had some time to clear the conflicts. Thanks again for your time reviewing the PR. [Stuck with 996 internship in China : ( ]

**Comment by drashna** [comment]
Sorry, there are some additional merge conflicts that have popped up.

**Comment by Codetector1374** [comment]
done

File: keyboards/ergodox_stm32/rules.mk (hunk 19 above)
Comment (drashna) [reply keyboards/ergodox_stm32/rules.mk]: And actually.... this should be added seperately:

```c
QUANTUM_LIB_SRC += i2c_master.c
```

**Comment by drashna** [comment]
Is the DFU bootloader's device ID and Vender ID known?

**Comment by drashna** [comment]
Ah, okay.   I asked for dfu args and dfu suffix settings, but that's not needed in this case.

**Comment by Codetector1374** [comment]
is this good to merge?

**Comment by drashna** [comment]
Looks good, thanks!
